### PR TITLE
Add option to package classes directory to a jar on the compile classpath

### PR DESCRIPTION
**Pull request by jjohannes**
A solution for #6784.

This PR requires #9670 to be merged first.

### How it works

Turn on using the `jar`, built by the _jar task_, instead of the class folder on the compile classpath by setting the following system property:

`org.gradle.java.compile-classpath-packaging=true`

### Measurements 
Results for measuring the [huge project](https://github.com/gradle/gradle/blob/5b7a84089e44e9eee446ee7401ff7f6fb4a44fda/subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/JavaTestProject.groovy#L38-L41) on Windows:

**Without compile-classpath-packaging** 

```
    Measuring Gradle tasks: clean classes
    Speed java-library-plugin: Results were inconclusive with 95% confidence.
    Difference: 22.622 s slower (22622 ms), 9.79%
      java-library-plugin median: 4.23 m min: 4.164 m, max: 4.266 m, se: 2.516 s}
      > [4.266 m, 4.23 m, 4.164 m]
      java-plugin median: 3.853 m min: 3.837 m, max: 3.866 m, se: 710.32 ms}
      > [3.837 m, 3.866 m, 3.853 m]
```

**With compile-classpath-packaging** 
```
    Measuring Gradle tasks: clean classes
    Speed java-library-plugin: Results were inconclusive with 95% confidence.
    Difference: 10.114 s faster (10114 ms), -4.37%
      java-library-plugin median: 3.688 m min: 3.671 m, max: 3.72 m, se: 1.22 s}
      > [3.688 m, 3.671 m, 3.72 m]
      java-plugin median: 3.856 m min: 3.839 m, max: 3.86 m, se: 544.53 ms}
      > [3.839 m, 3.86 m, 3.856 m]
```

### CODE REVIEW

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/KotlinDslFileContentGenerator.groovy [hunk 1]
```diff
@@ -31,6 +31,11 @@ class KotlinDslFileContentGenerator extends FileContentGenerator {
         'val noJavaLibraryPlugin = hasProperty("noJavaLibraryPlugin")'
     }
 
+    @Override
+    protected String javaPluginConfiguration() {
+        "extensions.getByType<JavaPluginExtension>().apply { packageClassesFromProjectDependencies() }"
```
Comment (eskatos) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/KotlinDslFileContentGenerator.groovy]:
```suggestion
        "configure<JavaPluginExtension> { packageClassesFromProjectDependencies() }"
```

**Comment by big-guy** [comment]
Isn't this just recreating the regular jar without resources?  Does the lack of resources explain why it's 10s faster than just the Java plugin?
If there are performance improvements that could be made to the Java plugin, the advice could just be to use the Java plugin when a project is over a certain size?

**Comment by big-guy** [comment]
Thanks @jjohannes.  I was just curious.

Would it make sense to just make this the way it works instead of using a property?  e.g., on Linux/mac, we always want classes-dir packaging and on Windows, we always want jar packaging because we know it's faster and just let variant dependency management pick the right variant?

One interesting wrinkle with this is how it affects the build cache.  Would two builds (one on Linux and one on Windows) with otherwise identical inputs (only difference is classes vs jars) be able to pull the other's result from the cache? If the answer is "no", then that's a step back from what they can do today, I think.

**Comment by ljacomet** [comment]
Closed by accident and could not re-open. See #9892

**Pull request closed without merging**